src/server.js

```diff
@@ -11,31 +12,33 @@ app.get('/', async (req, res, next) => {
   renderToString(<App environment={environment} />)
   const relayData = await environment.relaySSRMiddleware.getCache()
 
-  setTimeout(() => {
-    const html = renderToString(<App environment={environment} />)
+  const source = new RecordSource()
+  const store = new Store(source)
+  const html = renderToString(
+    <App
+      environment={new Environment({
+        network: Network.create(() => relayData[0][1]),
+        store,
+      })}
+    />
```

Comment:
`setTimeout` is used so that Relay has the ability to collect the data properly. I wasn't able to figure out a way to handle this otherwise. 

Why not pass in the environment returned from `const environment = createRelayEnvironment()`?

Reply:
I think the reason why Relay need to have `setTimeout` to collect the data is that `react-relay-network-modern` will make a `Promise` to get the data for Relay. This way make Relay need the time to get the data. As a result, I use `Network.create(() => relayData[0][1])` to make `fetchQuery` not be a `Promise`. Relay will not need to wait the time to get the data.

I did not read the code to find the real reason, but I had used this way to make SSR in the many times before I change use `react-relay-network-modern`. I think this change does not have any problem, now.

Reply:
So I pulled this down and it works well, but the issue is when working server-side only. By re-creating the network layer we loose things that relay-network-modern provides, which can be seen by disabling JavaScript on the client. If you can find a way to work with the existing network layer that's been created while also removing the `setTimeout` then I'd be happy to merge.

Reply:
I am not sure what you said. I disable JavaScripts, but this look like the same.
<img width="1275" alt="screen shot 2018-10-04 at 9 42 24 am" src="https://user-images.githubusercontent.com/8859179/46448519-d61e7e80-c7b9-11e8-8c08-8ad0601a0df3.png">

Can you tell more detail or give some screen shot to let me know?
I will try to fix your issue.

Reply:
Meaning, your solution via 

```js
Network.create(() => relayData[0][1]
```

works, and things SSR render, but then during subsequent server requests we lose the middlewares (such as logging) provided by `relay-network-modern` which are created created just above via

```js
const environment = createRelayEnvironment()
```

and passed into `<App environment={environment} />`. Instead an entirely new network is being created and passed in, which isn't `relay-network-modern`. 

If its possible to find a way to remove the setTimeout while preserving the existing network that's created here https://github.com/damassi/react-relay-network-modern-ssr-example/pull/3/files#diff-e6a5b42b2f7a26c840607370aed5301aR11 then that would be ideal.

Reply:
I will try to use `createRelayEnvironment` to render, but I have a question.

Most of `relay-network-modern` tools are used to build Relay network and Relay network is used to handle request the graphql data. All requests are handled in [here](https://github.com/damassi/react-relay-network-modern-ssr-example/pull/3/files#diff-e6a5b42b2f7a26c840607370aed5301aR12). Why [the second rendering](https://github.com/damassi/react-relay-network-modern-ssr-example/pull/3/files#diff-e6a5b42b2f7a26c840607370aed5301aR19) needs to use `relay-network-modern`? I change to use this way is that I think the second rendering should just return data. The second rendering should not handle any request.

Reply:
Ah, yeah, I totally agree this is strange... So in `relay-network-modern-ssr` there's this from the docs:

```js
  // Once the RelayEnvironment is instantiated, two App renders need to be made in
  // order to prepare data for hydration:

  // First, kick off Relay requests with an initial render
  ReactDOMServer.renderToString(<App relayEnvironment={relayEnvironment} />);

  // Second, await while all data were recieved from graphql server
  const relayData = await relayServerSSR.getCache();

  // Third, render the app a second time now that the Relay store has been primed
  // and send HTML and bootstrap data to the client for rehydration.
  const appHtml = ReactDOMServer.renderToString(<App relayEnvironment={relayEnvironment} />);
```

What happens is that after the first render the data is cached, and so in the second render the cached data is injected in and the request is skipped completely. This isn't ideal because there's still a whole tree of components that renders twice, but it suffices until a fix can be found.

Reply:
Well, I think rendering twice is needed before `QueryRenderer` is supported `SSR` directly. As I know, the first rendering will not render the whole tree of components. This just use `query` and `variables` to query.  This look like the first rendering just triggers the `fetch`.

The first rendering can use this to trigger request, and this also work `SSR`:
```js
ReactDOMServer.renderToString(
  <QueryRenderer
    query={query}
    variables={variables}
    environment={relayEnvironment}
    render={() => null}
  />
);
```

The docs in `relay-network-modern-ssr` is right. The second rendering skips the request, but it still use `async/await` in `relay-network-modern`. This can not be used in some framework. The best way is the new network in the second rendering or use `setTimeout`.

Reply:
@HsuTing - I'll bring up this thread in my work slack tomorrow and see what people say. If server-side middleware isn't needed then I agree that the new second network seems like a fine tradeoff for better compatability across the ecosystem; I just hesitate to update this repo because its technically just an example implementation rather than a part of the library. It might be worthwhile to make a note over at `relay-network-modern-ssr`, however.

Reply:
I agree. I will make an `issue` to `relay-network-modern-ssr` to take about this `pr`.

Reply:
[PR](https://github.com/relay-tools/react-relay-network-modern-ssr/pull/10) had merged. 😸

Reply:
Nice, I'll go ahead and merge then 👍 Thanks for digging into this!